From b7f2c874f0f62bcacd4deb94d04290e38b31d21b Mon Sep 17 00:00:00 2001
From: Jarod Wilson <jarod@redhat.com>
Date: Fri, 30 May 2014 12:41:55 -0400
Subject: [redhat] kernel.spec: make weak-updates actually work

Message-id: <1401453715-18059-1-git-send-email-jarod@redhat.com>
Patchwork-id: 81241
O-Subject: [RHEL7 PATCH] [redhat] kernel.spec: make weak-updates actually work
Bugzilla: 1101036
RH-Acked-by: Phillip Lougher <plougher@redhat.com>

Bugzilla: 1101036

Somehow, we neglected to notice that the weak-updates infrastructure never
made its way into the RHEL7 kernel. This restores the functionality, with
special thanks to Akemi Yagi of elrepo.org for discovering the problem and
an initial patch.

This is going to have to be added to z-stream as well, or kmod updates
built against the GA kernel won't work with any other 7.0.z kernel.

CC: Phillip Lougher <plougher@redhat.com>
CC: Don Howards <dhoward@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/redhat/kernel.spec.template b/redhat/kernel.spec.template
index 761b3d1..49c9bbb 100644
--- a/redhat/kernel.spec.template
+++ b/redhat/kernel.spec.template
@@ -860,6 +860,7 @@ BuildKernel() {
     # dirs for additional modules per module-init-tools, kbuild/modules.txt
     mkdir -p $RPM_BUILD_ROOT/lib/modules/$KernelVer/extra
     mkdir -p $RPM_BUILD_ROOT/lib/modules/$KernelVer/updates
+    mkdir -p $RPM_BUILD_ROOT/lib/modules/$KernelVer/weak-updates
     # first copy everything
     cp --parents `find  -type f -name "Makefile*" -o -name "Kconfig*"` $RPM_BUILD_ROOT/lib/modules/$KernelVer/build
     cp Module.symvers $RPM_BUILD_ROOT/lib/modules/$KernelVer/build
@@ -1263,6 +1264,10 @@ fi\
 %{expand:%%posttrans %{?1}}\
 %{_sbindir}/new-kernel-pkg --package kernel%{?-v:-%{-v*}} --mkinitrd --dracut --depmod --update %{KVRA}%{?-v:.%{-v*}} || exit $?\
 %{_sbindir}/new-kernel-pkg --package kernel%{?1:-%{1}} --rpmposttrans %{KVRA}%{?1:.%{1}} || exit $?\
+if [ -x %{_sbindir}/weak-modules ]\
+then\
+    %{_sbindir}/weak-modules --add-kernel %{KVRA}%{?1:.%{1}} || exit $?\
+fi\
 %{nil}
 
 #
@@ -1291,6 +1296,10 @@ fi}\
 %define kernel_variant_preun() \
 %{expand:%%preun %{?1}}\
 %{_sbindir}/new-kernel-pkg --rminitrd --rmmoddep --remove %{KVRA}%{?1:.%{1}} || exit $?\
+if [ -x %{_sbindir}/weak-modules ]\
+then\
+    %{_sbindir}/weak-modules --remove-kernel %{KVRA}%{?1:.%{1}} || exit $?\
+fi\
 %{nil}
 
 %kernel_variant_preun
@@ -1441,6 +1450,7 @@ fi
 /lib/modules/%{KVRA}%{?2:.%{2}}/source\
 /lib/modules/%{KVRA}%{?2:.%{2}}/extra\
 /lib/modules/%{KVRA}%{?2:.%{2}}/updates\
+/lib/modules/%{KVRA}%{?2:.%{2}}/weak-updates\
 %ifarch %{vdso_arches}\
 /lib/modules/%{KVRA}%{?2:.%{2}}/vdso\
 /etc/ld.so.conf.d/kernel-%{KVRA}%{?2:.%{2}}.conf\
-- 
1.7.1